### PR TITLE
Rollover for data streams

### PR DESCRIPTION
**Pull request by danhermann**
Adds support to the `_rollover` endpoint for data streams.

Relates to #53100.

### CODE REVIEW

**Comment by danhermann** [comment]
@elasticmachine update branch

**Review by martijnvg — COMMENTED** [review]
This looks great! Can you modify the `DataStreamIT#testBasicScenario()` test to also do a rollover?

File: server/src/main/java/org/elasticsearch/action/admin/indices/rollover/MetadataRolloverService.java [hunk 1]
```diff
@@ -75,17 +83,33 @@ private RolloverResult(String rolloverIndexName, String sourceIndexName, Cluster
         }
     }
 
-    public RolloverResult rolloverClusterState(ClusterState currentState, String aliasName, String newIndexName,
+    public RolloverResult rolloverClusterState(ClusterState currentState, String aliasOrDataStream, String newIndexName,
                                                CreateIndexRequest createIndexRequest, List<Condition<?>> metConditions,
                                                boolean silent) throws Exception {
+        validate(currentState.metadata(), aliasOrDataStream, newIndexName);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/indices/rollover/MetadataRolloverService.java:89]: I wonder if for data streams we should disallow applying aliases, mappings and settings to the new index? It feels odd that you cannot specify them on create data stream but can on rollover?

File: server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequestBuilder.java [hunk 2]
```diff
@@ -34,7 +34,7 @@ public RolloverRequestBuilder(ElasticsearchClient client, RolloverAction action)
     }
 
     public RolloverRequestBuilder setAlias(String alias) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequestBuilder.java:36]: I think we need to also rename this method?

**Review by henningandersen — COMMENTED** [review]
Thanks Dan, this looks great. I wonder if we could avoid the "alias or data stream" in both messages and variable/field names by calling this the "rollover target" instead? Could it make sense to hack a bit on the documentation to see if we can find a good name for this?

**Comment by danhermann** [comment]
@henningandersen, I've opened https://github.com/elastic/elasticsearch/pull/55551 to start on the docs for this.

**Comment by danhermann** [comment]
> This looks great! Can you modify the `DataStreamIT#testBasicScenario()` test to also do a rollover?

@martijnvg, I'll add a test to `DataStreamIT`. Should this have REST tests, too?

File: server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java [hunk 3]
```diff
@@ -99,14 +99,14 @@ protected void masterOperation(Task task, final RolloverRequest rolloverRequest,
                                    final ActionListener<RolloverResponse> listener) throws Exception {
         MetadataRolloverService.RolloverResult preResult =
             rolloverService.rolloverClusterState(state,
-                rolloverRequest.getAlias(), rolloverRequest.getNewIndexName(), rolloverRequest.getCreateIndexRequest(),
+                rolloverRequest.getRolloverTarget(), rolloverRequest.getNewIndexName(), rolloverRequest.getCreateIndexRequest(),
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:102]: The `TransportRolloverAction#checkBlock()` method should be changed.

The following

```
IndicesOptions indicesOptions = IndicesOptions.fromOptions(true, true,
            request.indicesOptions().expandWildcardsOpen(), request.indicesOptions().expandWildcardsClosed());
```

should be changed into:

```
IndicesOptions indicesOptions = IndicesOptions.fromOptions(true, true,
            request.indicesOptions().expandWildcardsOpen(), request.indicesOptions().expandWildcardsClosed(), 
            request.indicesOptions().expandWildcardsHidden(), request.indicesOptions().allowAliasesToMultipleIndices(), 
            request.indicesOptions().forbidClosedIndices(), request.indicesOptions().ignoreAliases(), 
            request.indicesOptions().ignoreThrottled(), request.indicesOptions().includeDataStreams());
```

This why the modified test in `DataStreamIT` failed.

**Review by martijnvg — APPROVED** [review]
Left one comment, LGTM otherwise.

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@henningandersen, I believe I have addressed all your review comments if you want to confirm that.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by danhermann** [comment]
Thanks, @henningandersen and @martijnvg!

**Pull request merged**